### [2TDSPX] DISRUPTIVE ARCHITECTURES: IOT, IOB E GENERATIVE AI

> * RM558948 - Allan Brito Moreira
> * RM558868 Caio Liang
> * RM98276 - Levi Magni

#### Checkpoint 02 - Treinamento de redes neurais com Keras (dados tabulares)

1. **Classificação Multiclasse**

2. **Regressão**


In [94]:
# Importando bibliotecas
import pandas as pd
import numpy as np

from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, mean_squared_error, r2_score
from sklearn.datasets import fetch_california_housing

from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical

 ### **Classificação Multiclasse**

 Dataset utilizado: [Wine dataset (UCI)](https://archive.ics.uci.edu/dataset/109/wine)

##### **1. Treinar uma rede neural em Keras para classificar vinhos em 3 classes.**
  - Configuração mínima: 2 camadas ocultas com 32 neurônios cada, função de ativação ReLU.
  - Camada de saída com 3 neurônios, função de ativação Softmax.
  - Função de perda: categorical_crossentropy.
  - Otimizador: Adam

In [28]:
# Carregar o conjunto de dados
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/wine/wine.data"
column_names = ['Class', 'Alcohol', 'Malic_acid', 'Ash', 'Alcalinity_of_ash', 'Magnesium',
                'Total_phenols', 'Flavanoids', 'Nonflavanoid_phenols', 'Proanthocyanins',
                'Color_intensity', 'Hue', 'OD280_OD315_of_diluted_wines', 'Proline']
data = pd.read_csv(url, header=None, names=column_names)


In [29]:
data.head()

Class  Alcohol  Malic_acid   Ash  Alcalinity_of_ash  Magnesium  \
0      1    14.23        1.71  2.43               15.6        127   
1      1    13.20        1.78  2.14               11.2        100   
2      1    13.16        2.36  2.67               18.6        101   
3      1    14.37        1.95  2.50               16.8        113   
4      1    13.24        2.59  2.87               21.0        118   

   Total_phenols  Flavanoids  Nonflavanoid_phenols  Proanthocyanins  \
0           2.80        3.06                  0.28             2.29   
1           2.65        2.76                  0.26             1.28   
2           2.80        3.24                  0.30             2.81   
3           3.85        3.49                  0.24             2.18   
4           2.80        2.69                  0.39             1.82   

   Color_intensity   Hue  OD280_OD315_of_diluted_wines  Proline  
0             5.64  1.04                          3.92     1065  
1             4.38  1.05                          3.40     1050  
2             5.68  1.03                          3.17     1185  
3             7.80  0.86                          3.45     1480  
4             4.32  1.04                          2.93      735

In [30]:
data.Class

0      1
1      1
2      1
3      1
4      1
      ..
173    3
174    3
175    3
176    3
177    3
Name: Class, Length: 178, dtype: int64

In [31]:
# Seleção das features:
X = data.iloc[:, 1:]

In [32]:
X.shape

(178, 13)

In [33]:
X.head()

Alcohol  Malic_acid   Ash  Alcalinity_of_ash  Magnesium  Total_phenols  \
0    14.23        1.71  2.43               15.6        127           2.80   
1    13.20        1.78  2.14               11.2        100           2.65   
2    13.16        2.36  2.67               18.6        101           2.80   
3    14.37        1.95  2.50               16.8        113           3.85   
4    13.24        2.59  2.87               21.0        118           2.80   

   Flavanoids  Nonflavanoid_phenols  Proanthocyanins  Color_intensity   Hue  \
0        3.06                  0.28             2.29             5.64  1.04   
1        2.76                  0.26             1.28             4.38  1.05   
2        3.24                  0.30             2.81             5.68  1.03   
3        3.49                  0.24             2.18             7.80  0.86   
4        2.69                  0.39             1.82             4.32  1.04   

   OD280_OD315_of_diluted_wines  Proline  
0                          3.92     1065  
1                          3.40     1050  
2                          3.17     1185  
3                          3.45     1480  
4                          2.93      735

In [34]:
# Dados da coluna target:
y = data['Class']

In [35]:
y

0      1
1      1
2      1
3      1
4      1
      ..
173    3
174    3
175    3
176    3
177    3
Name: Class, Length: 178, dtype: int64

In [36]:
# Converter as classes para começar do 0 (0, 1, 2) em vez de (1, 2, 3)
y = y - 1

In [37]:
# Usando o Label Encoder para converter colunas categóricas em numéricas
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(y)
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2])

In [44]:
# Separação de dados de treino e teste
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [45]:
# Normalização dos dados
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [46]:
# Converter labels para categorical (one-hot encoding)
y_train_categorical = to_categorical(y_train, num_classes=3)
y_test_categorical = to_categorical(y_test, num_classes=3)

In [47]:
# Construção do modelo
my_nn = Sequential()

In [48]:
X.shape[1]

13

In [49]:
# Criando primeira camada oculta
my_nn.add(Dense(32, input_dim=13, activation='relu'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [50]:
# Segunda camada oculta
my_nn.add(Dense(32, activation='relu'))

In [51]:
# Camada de saída (3 neurônios, função de ativação softmax)
my_nn.add(Dense(3, activation='softmax'))

In [52]:
my_nn.summary()

Model: "sequential_4"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_3 (Dense)                 │ (None, 32)             │           448 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 32)             │         1,056 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 3)              │            99 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 1,603 (6.26 KB)

 Trainable params: 1,603 (6.26 KB)

 Non-trainable params: 0 (0.00 B)

In [53]:
# Compilar o modelo
my_nn.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [54]:
# Treinamento do modelo
history = my_nn.fit(X_train_scaled, y_train_categorical, epochs=100, batch_size=32, validation_data=(X_test_scaled, y_test_categorical), verbose=1)


Epoch 1/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 1s 60ms/step - accuracy: 0.1009 - loss: 0.7909 - val_accuracy: 0.0833 - val_loss: 0.7654
Epoch 2/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 17ms/step - accuracy: 0.1864 - loss: 0.7340 - val_accuracy: 0.1667 - val_loss: 0.7209
Epoch 3/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 16ms/step - accuracy: 0.2915 - loss: 0.6906 - val_accuracy: 0.3056 - val_loss: 0.6833
Epoch 4/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 17ms/step - accuracy: 0.3979 - loss: 0.6538 - val_accuracy: 0.4444 - val_loss: 0.6491
Epoch 5/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 17ms/step - accuracy: 0.5499 - loss: 0.6200 - val_accuracy: 0.6111 - val_loss: 0.6178
Epoch 6/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 16ms/step - accuracy: 0.6405 - loss: 0.5898 - val_accuracy: 0.6667 - val_loss: 0.5882
Epoch 7/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 17ms/step - accuracy: 0.7217 - loss: 0.5565 - val_accuracy: 0.8333 - val_loss: 0.5588
Epoch 8/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 16ms/step - accuracy: 0.7992 - loss: 0.5188 - val_accuracy: 0.8611 - val_loss:

In [55]:
# Avaliação do modelo
test_loss, test_accuracy = my_nn.evaluate(X_test_scaled, y_test_categorical, verbose=0)
print(f"Acurácia no conjunto de teste: {test_accuracy:.4f}")

Acurácia no conjunto de teste: 1.0000


##### **2. Comparar os resultados com um modelo do scikit-learn (RandomForestClassifier).**

In [56]:
# Treinando modelo RandomForestClassifier
rf_model = RandomForestClassifier(n_estimators=100, random_state=42)
rf_model.fit(X_train_scaled, y_train)

RandomForestClassifier(random_state=42)

In [57]:
# Predições Random Forest
y_pred_rf = rf_model.predict(X_test_scaled)

# Avaliar o modelo
accuracy_rf = accuracy_score(y_test, y_pred_rf)
print(f"Acurácia do Random Forest: {accuracy_rf:.4f}")

Acurácia do Random Forest: 1.0000


In [86]:
# Comparar os resultados
print("Comparação de resultados:")
print(f"Acurácia da rede neural em Keras: {test_accuracy:.4f}")
print(f"Acurácia do RandomForestClassifier: {accuracy_rf:.4f}")

Comparação de resultados:
Acurácia da rede neural em Keras: 1.0000
Acurácia do RandomForestClassifier: 1.0000


##### **3. Registrar métricas de acurácia e discutir qual modelo teve melhor desempenho.**

  > Acurácia da rede neural em Keras: `1.0000`

  > Acurácia do RandomForestClassifier: `1.0000`

  Ambos os modelos (rede neural em Keras e RandomForestClassifier) alcançaram 100% de acurácia no conjunto de teste para este dataset específico. Isso indica que ambos os modelos foram capazes de classificar corretamente todas as instâncias no conjunto de teste.

### **Regressão**

Dataset utilizado: [California Housing dataset (scikit-learn)](https://scikit-learn.org/stable/datasets/real_world.html#california-housing-dataset)

1. Treinar uma rede neural em Keras para prever o valor médio das casas.
    - Configuração mínima: 3 camadas ocultas com 64, 32 e 16 neurônios, função de ativação ReLU.
    - Camada de saída com 1 neurônio, função de ativação Linear.
    - Função de perda: mse.
    - Otimizador: Adam.


In [61]:
# Carregar o conjunto de dados
housing = fetch_california_housing()
X_housing = pd.DataFrame(housing.data, columns=housing.feature_names)
y_housing = housing.target

In [62]:
X_housing.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  
0    -122.23  
1    -122.22  
2    -122.24  
3    -122.25  
4    -122.25

In [63]:
X_housing.shape

(20640, 8)

In [64]:
y_housing.shape

(20640,)

In [65]:
# Informações sobre o target
print("Valor médio das casas (target):")
print(f"Mínimo: ${y_housing.min():.2f}")
print(f"Máximo: ${y_housing.max():.2f}")
print(f"Média: ${y_housing.mean():.2f}")

Valor médio das casas (target):
Mínimo: $0.15
Máximo: $5.00
Média: $2.07


In [66]:
# Separação de dados de treino e teste
X_train_housing, X_test_housing, y_train_housing, y_test_housing = train_test_split(X_housing, y_housing, test_size=0.2, random_state=42)

In [67]:
X_train_housing.shape

(16512, 8)

In [68]:
X_test_housing.shape

(4128, 8)

In [69]:
# Normalização dos dados
scaler_housing = StandardScaler()
X_train_housing_scaled = scaler_housing.fit_transform(X_train_housing)
X_test_housing_scaled = scaler_housing.transform(X_test_housing)

In [70]:
# Construção do modelo
my_nn_regression = Sequential()

In [71]:
# 1ª camada oculta: 64 neurônios
my_nn_regression.add(Dense(64, input_dim=8, activation='relu'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [72]:
# 2ª camada oculta: 32 neurônios
my_nn_regression.add(Dense(32, activation='relu'))

In [73]:
# 3ª camada oculta: 16 neurônios
my_nn_regression.add(Dense(16, activation='relu'))

In [74]:
# Camada de saída com 1 neurônio, função de ativação Linear
my_nn_regression.add(Dense(1, activation='linear'))

In [76]:
my_nn_regression.summary()

Model: "sequential_5"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_6 (Dense)                 │ (None, 64)             │           576 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_7 (Dense)                 │ (None, 32)             │         2,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_8 (Dense)                 │ (None, 16)             │           528 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_9 (Dense)                 │ (None, 1)              │            17 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 3,201 (12.50 KB)

 Trainable params: 3,201 (12.50 KB)

 Non-trainable params: 0 (0.00 B)

In [75]:
# Compilar o modelo
my_nn_regression.compile(optimizer='adam', loss='mse', metrics=['mae'])

In [77]:
# Treinamento do modelo
history_regression = my_nn_regression.fit(X_train_housing_scaled, y_train_housing, epochs=100, batch_size=32,
                                          validation_data=(X_test_housing_scaled, y_test_housing), verbose=1)

Epoch 1/100
516/516 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - loss: 1.8440 - mae: 0.9491 - val_loss: 0.4440 - val_mae: 0.4596
Epoch 2/100
516/516 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 0.4143 - mae: 0.4560 - val_loss: 0.3802 - val_mae: 0.4339
Epoch 3/100
516/516 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 0.3664 - mae: 0.4292 - val_loss: 0.3613 - val_mae: 0.4172
Epoch 4/100
516/516 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 0.3576 - mae: 0.4234 - val_loss: 0.3394 - val_mae: 0.4000
Epoch 5/100
516/516 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 0.3436 - mae: 0.4125 - val_loss: 0.3241 - val_mae: 0.4007
Epoch 6/100
516/516 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 0.3288 - mae: 0.3970 - val_loss: 0.3230 - val_mae: 0.3998
Epoch 7/100
516/516 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 0.3156 - mae: 0.3929 - val_loss: 0.3121 - val_mae: 0.3837
Epoch 8/100
516/516 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 0.3141 - mae: 0.3879 - val_loss: 0.3061 - val_mae: 0.3884
Epoch 9/100
516/516 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/

In [80]:
# Avaliação da rede neural
test_loss_regression, test_mae_regression = my_nn_regression.evaluate(X_test_housing_scaled, y_test_housing, verbose=0)
print(f"MSE da rede neural: {test_loss_regression:.4f}")
print(f"MAE da rede neural: {test_mae_regression:.4f}")

MSE da rede neural: 0.2629
MAE da rede neural: 0.3345


In [89]:
# Predições da rede neural
y_pred_nn_regression = my_nn_regression.predict(X_test_housing_scaled)
r2_nn = r2_score(y_test_housing, y_pred_nn_regression)
print(f"R² da rede neural: {r2_nn:.4f}")

129/129 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step
R² da rede neural: 0.7994


##### **2. Comparar os resultados com um modelo do scikit-learn (LinearRegression).**

In [90]:
# Treinando modelo LinearRegression
lr_regression = LinearRegression()
lr_regression.fit(X_train_housing_scaled, y_train_housing)

LinearRegression()

In [91]:
# Predições LinearRegression
y_pred_lr_regression = lr_regression.predict(X_test_housing_scaled)
mse_lr = mean_squared_error(y_test_housing, y_pred_lr_regression)
r2_lr = r2_score(y_test_housing, y_pred_lr_regression)
print(f"MSE da LinearRegression: {mse_lr:.4f}")
print(f"R² da LinearRegression: {r2_lr:.4f}")

MSE da LinearRegression: 0.5559
R² da LinearRegression: 0.5758


In [98]:
# RMSE da rede neural
rmse_nn = np.sqrt(test_loss_regression)
print(f"RMSE da Rede Neural: {rmse_nn:.4f}")

RMSE da Rede Neural: 0.5127


In [99]:
# RMSE da LinearRegression
rmse_lr = np.sqrt(mse_lr)
print(f"RMSE da Linear Regression: {rmse_lr:.4f}")

RMSE da Linear Regression: 0.7456


In [101]:
# Comparar os resultados
print("Comparação de resultados:")
print(f"Acurácia da rede neural MSE: {test_loss_regression:.4f}, RMSE: {rmse_nn:.4f}, R²: {r2_nn:.4f}")
print(f"Acurácia do LinearRegression: {mse_lr:.4f}, RMSE: {rmse_lr:.4f}, R²: {r2_lr:.4f}")

Comparação de resultados:
Acurácia da rede neural MSE: 0.2629, RMSE: 0.5127, R²: 0.7994
Acurácia do LinearRegression: 0.5559, RMSE: 0.7456, R²: 0.5758


##### **3. Registrar métricas de erro (RMSE ou MAE) e discutir qual modelo teve melhor desempenho.**

**Métricas de erro:**

* Rede neural em Keras:
    * MSE: `0.2629`
    * RMSE: `0.5127`
    * MAE: `0.3345`
    * R²: `0.7994`

* LinearRegression (scikit-learn):
    * MSE: `0.5559`
    * RMSE: `0.7456`
    * MAE: `0.5955`
    * R²: `0.5758`

**Desempenho:**

Ao comparar as métricas de erro para os modelos de regressão, podemos observar que a rede neural em Keras apresentou um desempenho significativamente melhor do que o modelo de LinearRegression.

* **MSE (Mean Squared Error) e RMSE (Root Mean Squared Error):**
  * Tanto o MSE quanto o RMSE da rede neural (`0.2629` e `0.5127`, respectivamente) são consideravelmente menores do que os da LinearRegression (`0.5559` e `0.7456`). Esses métricas penalizam erros maiores, e os valores mais baixos para a rede neural indicam que suas previsões estão, em média, mais próximas dos valores reais e com menos erros grandes.

* **MAE (Mean Absolute Error):**
  * O MAE da rede neural (`0.3345`) também é menor do que o da LinearRegression (`0.5955`). O MAE fornece uma medida da magnitude média dos erros, e o valor menor para a rede neural confirma sua maior precisão nas previsões.

* **R² (Coefficient of Determination):**
  * O valor de R² para a rede neural (`0.7994`) é substancialmente mais alto do que o da LinearRegression (`0.5758`). O R² indica a proporção da variância na variável dependente que é previsível a partir das variáveis independentes. Um R² mais próximo de 1 indica que o modelo explica uma maior proporção da variabilidade dos dados.

Com base nessas métricas, a rede neural em Keras demonstrou ter um desempenho superior na tarefa de prever o valor médio das casas neste dataset, com erros de previsão menores e uma maior capacidade de explicar a variância nos dados.